File: python/other/dp/coin-change.py
class Solution:
    def coin_change(self, coins, target):
        dp = [target + 1 for _ in range(target + 1)]
        dp[0] = 0
        for i in range(1, target + 1):
            for c in coins:
                if c <= i:
                    dp[i] = min(dp[i], dp[i - c] + 1)
        return dp[target]


if __name__ == '__main__':
    s = Solution()
    print(s.coin_change([1, 2, 5], 11))

File: code/java/src/main/java/cn/az/code/practices/dp/CoinChange.java
package cn.az.code.practices.dp;

import java.util.Arrays;

/**
 * little-code
 *
 * @author azusachino
 * @since 27/9/2020 23:20
 */
public class CoinChange {

    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);

        // base case
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int c : coins) {
                if (i > c) {
                    dp[i] = Math.min(dp[i], dp[i - c] + 1);
                }
            }
        }
        return dp[amount] == amount + 1 ? -1 : dp[amount];
    }
}
